### PR TITLE
Add LD7032 support to QP.

### PR DESCRIPTION
**Pull request by daskygit**
## Description

Depends on https://github.com/qmk/qmk_firmware/pull/19997

I've tested I2C using a 96x24 display.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by daskygit** [comment]
It's pretty much ready to go minus the poor i2c/spi comms handling. Opened so duplicate work doesn't get done (although unlikely).

**Comment by daskygit** [comment]
best I sort the documentation conflicts

File: drivers/painter/ld7032/qp_ld7032.c [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+
+    qp_comms_bulk_command_sequence(device, ld7032_init_sequence, sizeof(ld7032_init_sequence));
+
+    uint8_t display_y_start = 40 - driver->oled.base.panel_height;
+    uint8_t display_x_start = (128 - driver->oled.base.panel_width) / 2;
+
+    uint8_t ld7032_memory_setup[] = {
+        // Command,                    Delay,  N, Data[N]
+        LD7032_DISP_SIZE_X, 0, 2, 0x00, 0x7F, LD7032_DISP_SIZE_Y, 0, 2, 0x00, 0x27, LD7032_X_DISP_START, 0, 1, 0x0, LD7032_Y_DISP_START, 0, 1, 0x0,
+    };
```
Comment (tzarc) [inline drivers/painter/ld7032/qp_ld7032.c:213]: // clang-format off
    uint8_t ld7032_memory_setup[] = {
        // Command,        Delay,  N, Data[N]
        LD7032_DISP_SIZE_X,     0, 2, 0x00, 0x7F,
        LD7032_DISP_SIZE_Y,     0, 2, 0x00, 0x27,
        LD7032_X_DISP_START,    0, 1, 0x0,
        LD7032_Y_DISP_START,    0, 1, 0x0,
    };
    // clang-format on

**Review by tzarc — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

File: docs/quantum_painter.md [hunk 2]
```diff
@@ -478,6 +480,40 @@ Native color format mono2 is compatible with SH1106
...
+QUANTUM_PAINTER_DRIVERS += ld7032_i2c
+```
+
+Creating a SH1106 device in firmware can then be done with the following APIs:
+
+```c
+// SPI-based SH1106:
+painter_device_t qp_ld7032_make_spi_device(uint16_t panel_width, uint16_t panel_height, pin_t chip_select_pin, pin_t dc_pin, pin_t reset_pin, uint16_t spi_divisor, int spi_mode);
+// I2C-based SH1106:
+painter_device_t qp_ld7032_make_i2c_device(uint16_t panel_width, uint16_t panel_height, uint8_t i2c_address);
```
Comment (sigprof) [inline docs/quantum_painter.md:501]:
```suggestion
// SPI-based LD7032:
painter_device_t qp_ld7032_make_spi_device(uint16_t panel_width, uint16_t panel_height, pin_t chip_select_pin, pin_t dc_pin, pin_t reset_pin, uint16_t spi_divisor, int spi_mode);
// I2C-based LD7032:
painter_device_t qp_ld7032_make_i2c_device(uint16_t panel_width, uint16_t panel_height, uint8_t i2c_address);
```

File: docs/quantum_painter.md [hunk 3]
```diff
@@ -478,6 +480,40 @@ Native color format mono2 is compatible with SH1106
...
+The maximum number of displays of each type can be configured by changing the following in your `config.h` (default is 1):
+
+```c
+// 3 SPI displays:
+#define LD7032_NUM_SPI_DEVICES 3
+// 3 I2C displays:
+#define LD7032_NUM_I2C_DEVICES 3
+```
+
+Native color format mono2 is compatible with LD7032
```
Comment (sigprof) [inline docs/quantum_painter.md:515]:
```suggestion
Native color format mono2 is compatible with LD7032.
```

**Review by sigprof — CHANGES_REQUESTED** [review]

**Review by sigprof — CHANGES_REQUESTED** [review]

File: drivers/painter/ld7032/qp_ld7032.c [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+    uint8_t y_view_offset = driver->offset_y;
+
+    for (int y_pos = y_start; y_pos <= y_end; y_pos++) {
+        int y_new_pos = y_pos;
+        if (inverted) {
+            y_new_pos = y_end - y_pos;
+        }
+        uint8_t packet[x_length];
+        memset(packet, 0, sizeof(packet));
+        memcpy(packet, &framebuffer[(y_pos * (driver->panel_width >> 3)) + x_start], x_length);
```
[the remark this answers is not in the document]
Comment (daskygit) [reply drivers/painter/ld7032/qp_ld7032.c:98]: Ta, I'd be a little surprised if the driver is ever used for anything else. It does seem like there's potential for another steelseries mouse but looks like it's the same display size.

[the remark this answers is not in the document]
Comment (sigprof) [reply drivers/painter/ld7032/qp_ld7032.c]: Yeah, the only thing with LD7032 that I was able to find on AliExpress is https://www.aliexpress.com/item/1005006897048850.html, and the description says that the model is off production (it also uses an unusual FPC connector with 0.3mm pitch and two rows of contacts).

**Review by tzarc — APPROVED** [review]

**Comment by daskygit** [comment]
Ta

**Pull request merged**